Add border on Title when focused

pinarol opened the pull request:
Fix https://github.com/wordpress-mobile/gutenberg-mobile/issues/540
Fix https://github.com/wordpress-mobile/gutenberg-mobile/issues/551

This change adds borders to the Title when focused in order to match other blocks.
And also fixes a slight shift happening on Android when we focus to a block.

gutenberg PR needs to be merged first: https://github.com/WordPress/gutenberg/pull/13970

![android-title-border](https://user-images.githubusercontent.com/5032900/53090362-3a02d300-3520-11e9-8b0f-97a27a110d63.gif)

![ios-title-border](https://user-images.githubusercontent.com/5032900/53090388-53a41a80-3520-11e9-9e6f-3aedcf037397.gif)

**To Test**

Testing prerequisites

For WPiOS

Checkout the PRs branch to any arbitrary folder and cd .. to it
run yarn install, yarn start
Open XCode WPiOS on the latest develop
Clean build folder on Xcode, and then run the app

For WPAndroid

open grade.properties at WordPress-Android folder
add wp.BUILD_GUTENBERG_FROM_SOURCE = true to grade.properties
checkout the PRs branch in the subrepo of WordPress-Android repo
cd to WordPress-Android/libs/gutenberg-mobile
run yarn install, yarn start
yarn wpandroid on a separate terminal in the same directory

Test Steps:

WPiOS:

- Open a post and tap the title
- Verify you see focus indicating borders around the title just like the other blocks
- Verify there's no slight shifting on portrait mode when we focus on a block because there should be no left right borders

iPhone X and iPad special case:
- Verify that you do see the left right borders when there's empty space around the block like there's on iPhone X landscape mode, and iPad portrait and landscape modes.

WPAndroid:

- Open a post and tap the title
- Verify you see focus indicating borders around the title just like the other blocks
- Verify that the slight shift that is described here is fixed also: https://github.com/wordpress-mobile/gutenberg-mobile/issues/551

iamthomasbishop commented on src/block-management/block-manager.scss:
```diff
@@ -56,4 +57,20 @@
+}
+
+.blockHolderFocusedSemiBordered {
+	border-color: $gray-lighten-30;
```
Would it make sense to use css `box-shadow` to apply these focus states? In theory that should prevent any shifting, just applying a shadow on the outside edge of the element.

pinarol replied:
Unfortunately box-shadow property is not available in React Native and as I investigated it is not a very straight forward change to replace borders with shadows in RN. Because Android and iOS have very different methods for doing it and both methods have different side effects that needs to be addressed.

iOS has these props:
https://facebook.github.io/react-native/docs/shadow-props#shadowcolor

Android has the elevation option but there are no other options to fine-tune the shadow effect. So I am not even sure it'll give the results we want.
https://facebook.github.io/react-native/docs/view-style-props#elevation

I tried to apply these quickly but it messed up the child views' styles in different ways, so it didn't look like an ideal last minute change, but, if you want we can open a separate issue for this and try later? 

In this PR, I fixed the shifting of Android blocks. I don't expect any shifting will happen in cases where the block width is equal to window width, because in that case we don't have left&right borders at all.

etoledom commented on src/block-management/block-manager.js:
```diff
@@ -201,6 +226,7 @@ export class BlockManager extends React.Component<PropsType, StateType> {
 					keyboardShouldPersistTaps="always"
 					style={ styles.list }
 					data={ this.props.blockClientIds }
+					extraData={ [ this.state.isTitleFocused, this.state.isFullyBordered ] }
```
I don't see this property being used in `KeyboardAwareFlatList` or somewhere else. What is it for? 🤔

pinarol replied:
By passing extraData we make sure FlatList will re-render when extraData changes, otherwise changes won't get effect. If you look inside `keyboard-aware-flat-list.ios.js` or `keyboard-aware-flat-list.android.js` you can see all props are propagated to react-native's FlatList.

etoledom replied:
Perfect! Thank you.

pinarol commented:
@etoledom 👋 thanks for the review. I am going to address that issue about the empty post but before that I found another issue unfortunately, it looks a bit annoying and still couldn't come up with a solution. On Android, the title focus is getting lost in the first tap:

![android-title-focus-lost-on-first-tap](https://user-images.githubusercontent.com/5032900/53151270-d3cf8c00-35c3-11e9-8d9b-0b52a8b2c36d.gif)

I am guessing the re-render of the title view in order to add the borders might be somehow triggering a blur event on Android side, I've seen the blur event coming from Android side but not sure about the root cause of it. This is not happening on iOS. Still looking into it.

pinarol commented:
Hey @etoledom 👋  This is ready for another look. 

It looks like Android can somehow re-renders the title if we add border to it and this makes a blur event triggered. So I changed the way we add borders, I updated the containers views with borders with transparent color. When we focus I just change the border color. So we will not only get rid of unnecessary re-renders but also we are removing possibilities of the shifting effects when we focus. This is how the shadow effect would work also, cc @iamthomasbishop 

![ios-title-focus-landscape](https://user-images.githubusercontent.com/5032900/53169295-e3fc6100-35ed-11e9-9c69-43baf5e2c7cc.gif)

![android-titlefocus](https://user-images.githubusercontent.com/5032900/53169296-e3fc6100-35ed-11e9-88d9-7b992bd85d13.gif)

etoledom commented on src/block-management/block-manager.js:
```diff
@@ -172,21 +188,21 @@ export class BlockManager extends React.Component<PropsType, StateType> {
-					placeholder={ __( 'Add title' ) } />
-			</View>
+			<PostTitle
+				onRef={ ( ref ) => {
```
Nitpick: Can we name this `ref` instead of `onRef`? Since it seems to be more standard in React.

pinarol replied:
done with 67b509a
